File: iOS/SmartCampusApp/SmartCampus/Vertex.m
//
//  Vertex.m
//  SmartCampus
//
// Copyright (c) 2014, Aalborg University
// All rights reserved.

// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//     * Neither the name of the <organization> nor the
//       names of its contributors may be used to endorse or promote products
//       derived from this software without specific prior written permission.

// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL AAlBORG UNIVERSITY BE LIABLE FOR ANY
// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#import "Vertex.h"
#import "AggregateLocation.h"
#import "WifiMeasurement.h"
#import "Edge.h"

@implementation Vertex

@synthesize title = _title;

+ (Vertex *)vertexWithJson:(NSDictionary *)data {
    int vertexId = [[data valueForKey:@"ID"] intValue];
    AggregateLocation *location = [AggregateLocation new];

    NSArray *absolutes = [data valueForKey:@"AbsoluteLocations"];
    NSArray *symbolics = [data valueForKey:@"SymbolicLocations"];

    if ([absolutes count] == 1) {
        id absoluteData = [absolutes objectAtIndex:0];
        location.absolute = [AbsoluteLocation locationWithJson:absoluteData];
    }

    if ([symbolics count] == 1) {
        id symbolicData = [symbolics objectAtIndex:0];
        location.symbolic = [SymbolicLocation locationWithJson:symbolicData];
    }

    Vertex *vertex = [[Vertex alloc] initWithId:vertexId andLocation:location];
    vertex.buildingId = [[data valueForKey:@"Building_ID"] intValue];

    return vertex;
}

- (NSDictionary *)toJSON {
    NSDictionary *data = [[NSDictionary alloc] initWithObjectsAndKeys:
                          [[NSNumber numberWithInt:self.id] stringValue], @"ID",
                          [[NSNumber numberWithInt:self.buildingId] stringValue], @"Building_ID",
                          [NSArray arrayWithObject:[self.location.absolute toJSON]], @"AbsoluteLocations",
                          [NSArray arrayWithObject:[self.location.symbolic toJSON]], @"SymbolicLocations",
                          nil];

    return data;
}

- (id)initWithId:(int)id andAbsoluteLocation:(AbsoluteLocation *)absolute {
    return [self initWithId:id andLocation:[[AggregateLocation alloc] initWithAbsolute:absolute]];
}

- (id)initWithId:(int)id andLocation:(AggregateLocation *)location {
    if (self = [super init]) {
        self.id = id;
        self.location = location;
        self.inEdges = [NSMutableArray new];
        self.outEdges = [NSMutableArray new];
        self.fingerprints = [NSMutableArray new];
        self.radiusVertices = [NSMutableArray new];
    }

    return self;
}

- (id)initWithVertex:(Vertex *)vertex {
    if (self = [super init]) {
        self.id = vertex.id;
        self.location = [[AggregateLocation alloc] initWithAbsolute:[vertex.location.absolute copy]];
        self.inEdges = [vertex.inEdges copy];
        self.outEdges = [vertex.outEdges copy];
        self.fingerprints = [vertex.fingerprints copy];
        self.radiusVertices = [vertex.radiusVertices copy];
    }
    
    return self;
}

- (BOOL)addInEdge:(Edge *)edge {
    if (edge == nil || [self.inEdges containsObject:edge]) return NO;

    self.stairEndpoint = edge.staircase;
    self.elevatorEndpoint = edge.elevator;

    [self.inEdges addObject:edge];
    edge.destination = self;

    return YES;
}

- (BOOL)addInEdgeFromOrigin:(Vertex *)origin {
    if (origin == nil) return NO;

    [self addInEdge:[[Edge alloc] initWithOrigin:origin andDestination:self]];

    return YES;
}

- (BOOL)removeInEdge:(Vertex *)origin {
    Edge *edge = [[Edge alloc] initWithOrigin:origin andDestination:self];
    if (![self.inEdges containsObject:edge]) return NO;

    [self.inEdges removeObject:edge];

    return YES;
}

- (BOOL)addOutEdge:(Edge *)edge {
    if (edge == nil || [self.outEdges containsObject:edge]) return NO;

    self.stairEndpoint = edge.staircase;
    self.elevatorEndpoint = edge.elevator;

    [self.outEdges addObject:edge];
    edge.origin = self;

    return YES;
}

- (BOOL)addOutEdgeToDestination:(Vertex *)destination {
    if (destination == nil) return NO;

    [self addOutEdge:[[Edge alloc] initWithOrigin:self andDestination:destination]];

    return YES;
}

- (BOOL)removeOutEdge:(Vertex *)destination {
    Edge *edge = [[Edge alloc] initWithOrigin:self andDestination:destination];
    if (![self.outEdges containsObject:edge]) return NO;

    [self.outEdges removeObject:edge];

    return YES;
}

- (BOOL)hasFingerprints {
    return self.fingerprints != nil && [self numFingerprints] > 0;
}

- (int)numFingerprints {
    return [self.fingerprints count];
}

- (BOOL)addFingerprint:(WifiMeasurement *)measurement {
    if ([self hasFingerprints] || measurement == nil) return NO;
    
    [self.fingerprints addObject:measurement];
    measurement.vertex = self;

    return YES;
}

- (BOOL)removeFingerprint:(WifiMeasurement *)measurement {
    if (![self.fingerprints containsObject:measurement]) return NO;

    [self.fingerprints removeObject:measurement];

    return YES;
}

- (BOOL)addRadiusVertex:(Vertex *)vertex {
    for (Vertex *object in [self adjacentVertices]) {
        if ([object isEqual:vertex]) return NO;
    }

    [self.radiusVertices addObject:vertex];

    return YES;
}

- (BOOL)removeRadiusVertex:(Vertex *)vertex {
    if (![self.radiusVertices containsObject:vertex]) return NO;

    [self.radiusVertices removeObject:vertex];

    return YES;
}

- (NSArray *)adjacentVertices {
    NSMutableArray *adjacent = [NSMutableArray new];

    [adjacent addObjectsFromArray:[self opposingVertices:self.inEdges]];
    [adjacent addObjectsFromArray:[self opposingVertices:self.outEdges]];

    return adjacent;
}

- (NSArray *)opposingVertices:(NSArray *)edges {
    NSMutableArray *opposing = [NSMutableArray new];

    for (Edge *edge in edges) {
        [opposing addObject:[edge oppositeVertexOfVertex:self]];
    }

    return opposing;
}

- (BOOL)hasEdge:(Edge *)edge {
    return [self hasInEdge:edge] || [self hasOutEdge:edge];
}

- (BOOL)hasInEdge:(Edge *)edge {
    return [self.inEdges containsObject:edge];
}

- (BOOL)hasOutEdge:(Edge *)edge {
    return [self.outEdges containsObject:edge];
}

- (int)degree {
    return [self inDegree] + [self outDegree];
}

- (int)inDegree {
    return [self.inEdges count];
}

- (int)outDegree {
    return [self.outEdges count];
}

- (NSArray *)origins {
    return [self opposingVertices:self.inEdges];
}

- (NSArray *)destinations {
    return [self opposingVertices:self.outEdges];
}

- (BOOL)isEqual:(id)object {
    if (object == self) return YES;
    if (!object || ![object isKindOfClass:[self class]]) return NO;

    Vertex *other = (Vertex *) object;
    return [self.location isEqual:other.location];
}

- (int)hashCode {
    return [self.location hashcode];
}    

- (NSArray *)incidentEdges {
    NSMutableArray *incident = [NSMutableArray new];

    [incident addObjectsFromArray:self.inEdges];
    [incident addObjectsFromArray:self.outEdges];

    return incident;
}

- (double)latitude {
    return self.location.absolute.latitude;
}

- (double)longitude {
    return self.location.absolute.longitude;
}

- (double)altitude {
    return self.location.absolute.altitude;
}

- (NSString *)description {
    AbsoluteLocation* absolute = self.location.absolute;
    return [NSString stringWithFormat:@"(%g, %g, %g)",
            absolute.latitude, absolute.longitude, absolute.altitude];
}

#pragma mark TablePickerItem
- (NSString *)title {
    if (_title == nil) {
        AggregateLocation *location = self.location;

        if (location.symbolic != nil) {
            _title = location.symbolic.title;
        } else {
            _title = [NSString stringWithFormat:@"%g, %g",
                      location.absolute.latitude, location.absolute.longitude];
        }
    }

    return _title;
}

@end

File: WinPhone/SmartCampusLibrary/indoormodel/graph/Vertex.cs
/*
Copyright (c) 2014, Aalborg University
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of the <organization> nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL AAlBORG UNIVERSITY BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using com.smartcampus.indoormodel;
using com.smartcampus.wifi;

namespace com.smartcampus.indoormodel.graph
{
	public class Vertex
	{    
		private List<Edge> inEdges = new List<Edge>();
		private List<Edge> outEdges = new List<Edge>();
	
		//id (used for serialization on WinMobile - maybe not needed here)
		private int id;
	
		private AggregateLocation location = new AggregateLocation();
		//The fingerprints attached to the location
		private List<WifiMeasurement> fingerprints = new List<WifiMeasurement>();
		//Cf. "the graph paper" regarding the use of radius vertices
		private HashSet<Vertex> radiusVertices = new HashSet<Vertex>();
  
		private bool mIsStairEndpoint;
		private bool mIsElevatorEndpoint;
		
		public static int NEXT_ID = 0;
		public Vertex()
		{
		
		}
	
		public Vertex(int id, AbsoluteLocation absoluteLocation)
		{
			this.id = id;
			this.location = new AggregateLocation(absoluteLocation);
		}
	
		public Vertex(int id, AggregateLocation aggregateLocation)
		{
			this.id = id;
			this.location = aggregateLocation;
		}
	
		// Copy constructor
		public Vertex(Vertex aVertex) {
		
			double latitude = aVertex.getLocation().getAbsoluteLocation().getLatitude();
			double longitude = aVertex.getLocation().getAbsoluteLocation().getLongitude();
			double altitude = aVertex.getLocation().getAbsoluteLocation().getAltitude();
		
			AggregateLocation aggLoc = new AggregateLocation(new AbsoluteLocation(latitude, longitude, altitude));
			this.location = aggLoc;
			this.inEdges = aVertex.getInEdges();
			this.outEdges = aVertex.getOutEdges();
			this.id = aVertex.getId();
			this.fingerprints = aVertex.getFingerPrints();
			this.radiusVertices = aVertex.getRadiusVertices();
		}
	
		public bool addFingerprint(WifiMeasurement value)
		{
			if (fingerprints == null)
				return false;
			if (value == null)
				return false;
		
			fingerprints.Add(value);
			return true;
		}
	
		//Adds an in-edge, and sets the destination of the edge to this vertex
		public bool addInEdge(Edge e)
		{
			if (e == null)
				return false;
		
            e.setDestination(this);
			if (e.isElevator())
				this.setIsElevatorEndpoint(true);
			if (e.isStair())
				this.setIsStairEndpoint(true);
		
			if (inEdges.Contains(e))
				return false;
			
            inEdges.Add(e);
            return true;
		}    
	
		public bool addInEdge(Vertex origin)
		{
			if (origin == null)
				return false;
		
			Edge e = new Edge(origin, this);
			return addInEdge(e);    	
		}
	
		//Adds an out edge, and sets the origin explicitly to this vertex
		public bool addOutEdge(Edge e)
		{
			if (e == null)
				return false;
		
			e.setOrigin(this);
			if (e.isElevator())
				this.setIsElevatorEndpoint(true);
			if (e.isStair())
				this.setIsStairEndpoint(true);
		
			if (outEdges.Contains(e))
				return false;
			
            outEdges.Add(e);
            return true;
		}
	
		//Out- and in-edges must be in sync
		public bool addOutEdge(Vertex destination)
		{
			if (destination == null)
				return false;
		
			Edge e = new Edge(this, destination);
			return addOutEdge(e);
		}
	
		public bool addRadiusVertex(Vertex value)
		{
			//a radius vertex must not be present in the adjacent vertices
			foreach (Vertex v in this.adjacentVertices())
				if (v.Equals(value))
					return false;
		
			return radiusVertices.Add(value);
		}
	 
		public List<Vertex> adjacentVertices()
		{
			List<Vertex> result = new List<Vertex>();
			result.AddRange(opposites(inEdges));
			result.AddRange(opposites(outEdges));
			return result;
		}
	
		public bool containsEdge(Edge value)
		{
			return inEdges.Contains(value) || outEdges.Contains(value);
		}
	
		public int degree()
		{
			return inEdges.Count + outEdges.Count;
		}
	
		/**
		 * @return All vertices that this vertex is pointing to.
		 */
		public List<Vertex> destinations()
		{
			return opposites(outEdges);
		}
	
		/// <summary>
		///We compare vertices by their location, cf. Location.equals()
		/// </summary>
		/// <param name="other"></param>
		/// <returns>True, if the two vertices reside at the same (unique) location</returns>
		public override bool Equals(Object other)
		{
			if (other == null)
				return false;
		
			if (!(other is Vertex))
				return false;
		
			AggregateLocation thisLocation = this.getLocation();
			AggregateLocation otherLocation = ((Vertex)other).getLocation();
			if (thisLocation == null || otherLocation == null)
				return false;
			return thisLocation.Equals(otherLocation);
		}
		//Wifi measurements taken at this particular location
		public List<WifiMeasurement> getFingerPrints()
		{
			return fingerprints;
		}
	
		public int getId()
		{
			return this.id;
		}        
	
		public List<Edge> getInEdges()
		{
			return inEdges;
		}
	
		public AggregateLocation getLocation()
		{
			return location;
		}
	
		public int getNumFingerprints()
		{
			return fingerprints.Count;
		}    
	 
		public List<Edge> getOutEdges()
		{
			return outEdges;
		}
	
		public HashSet<Vertex> getRadiusVertices()
		{
			return radiusVertices;        
		}
	
		public override int GetHashCode()
		{
			return this.getLocation().GetHashCode();
		}    
	
		public List<Edge> incidentEdges()
		{
			List<Edge> result = new List<Edge>();
			result.AddRange(inEdges);
			result.AddRange(outEdges);
			return result;
		}
	
		//The remaining methods pertain to a vertex's position use. 
		//We don't bother to create a subclass as our vertex has a singular purpose in our application

		public int inDegree()
		{
			return inEdges.Count;
		}
		  
		public bool isElevatorEndpoint()
		{
			return this.mIsElevatorEndpoint;
		}
	
		public bool isStairEndpoint()
		{
			return mIsStairEndpoint;
		}
	
		//Helper method for origins(), destinations(), and adjacentVertices()
		private List<Vertex> opposites(IEnumerable<Edge> edges)
		{
			List<Vertex> result = new List<Vertex>();
			foreach (Edge e in edges)
				result.Add(e.Opposite(this));
			return result;
		}
	
		/**
		 * @return All vertices that have an outgoing edge to this vertex
		 */
		public List<Vertex> origins()
		{
			return opposites(inEdges);
		}
	
		public int outDegree()
		{
			return outEdges.Count;
		}
	
		public bool removeFingerprint(WifiMeasurement value)
		{
			if (!fingerprints.Contains(value))
				return false;
			return fingerprints.Remove(value);
		}
	
		public bool removeInEdge(Vertex origin)
		{
			Edge e = new Edge(origin, this);
			if (!inEdges.Contains(e))
				return false;
			return inEdges.Remove(e);
		}    	 
	
		public bool removeOutEdge(Vertex destination)
		{
			Edge e = new Edge(this, destination);
			if (!outEdges.Contains(e))
				return false;
			return outEdges.Remove(e);
		}
	
		public bool RemoveRadiusVertex(Vertex value)
		{
			return radiusVertices.Remove(value);
		}
	
		public void setId(int value)
		{
			this.id = value;
		}
	
		public void setIsElevatorEndpoint(bool val)
		{
			this.mIsElevatorEndpoint = val;
		}
	
		public void setIsStairEndpoint(bool val)
		{
			this.mIsStairEndpoint = val;
		}
	}
}
